core/src/main/java/edu/wpi/grip/core/sources/CameraSource.java

```diff
@@ -125,114 +124,120 @@ public void createFromProperties(EventBus eventBus, Properties properties) throw
-        if (frameThread.isPresent()) {
-            final Thread ex = frameThread.get();
-            ex.interrupt();
-            try {
-                ex.join(TimeUnit.SECONDS.toMillis(2));
-                if (ex.isAlive()) {
-                    throw new TimeoutException("Unable to terminate video feed from Web Camera");
-                }
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                //TODO: Move this into a logging framework
-                System.out.println("Caught Exception:");
-                e.printStackTrace();
-            } finally {
-                frameThread = Optional.empty();
-                // This will always run even if a timeout exception occurs
+    public void stop() throws TimeoutException, IllegalStateException {
+        synchronized (this) {
+            if (frameThread.isPresent()) {
+                final Thread ex = frameThread.get();
+                ex.interrupt();
                 try {
-                    grabber.ifPresent(grabber -> {
-                        try {
-                            grabber.stop();
-                        } catch (FrameGrabber.Exception e) {
-                            throw new IllegalStateException("A problem occurred trying to stop the frame grabber", e);
-                        }
-                    });
+                    ex.join(TimeUnit.SECONDS.toMillis(500));
+                    if (ex.isAlive()) {
+                        throw new TimeoutException("Unable to terminate video feed from Web Camera");
+                    }
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    //TODO: Move this into a logging framework
+                    System.out.println("Caught Exception:");
+                    e.printStackTrace();
                 } finally {
-                    // This will always run even if we fail to stop the grabber
-                    grabber = Optional.empty();
+                    // Clean up this resource as you can't restart a stopped thread
+                    this.frameThread = Optional.empty();
+                    // This will always run even if a timeout exception occurs
+                    try {
+                        grabber.stop();
+                    } catch (FrameGrabber.Exception e) {
+                        throw new IllegalStateException("A problem occurred trying to stop the frame grabber", e);
+                    }
                 }
+            } else {
+                throw new IllegalStateException("Tried to stop a Webcam that is already stopped.");
             }
-        } else {
-            throw new IllegalStateException("Tried to stop a Webcam that is already stopped.");
+        }
+        eventBus.post(new SourceStoppedEvent(this));
+        frameRateOutputSocket.setValue(0);
+    }
+
+    @Override
+    public boolean isRunning() {
+        synchronized (this) {
```

Comment:
I'm not sure I understand the difference between the two, please explain if there is one.